### PR TITLE
Add support for .jnilib extension on Mac OS

### PR DESCRIPTION
**Pull request by fengxue-IS**
- attempt to load native library with .jnilib extension
  after failed to load using .dylib extension

Close: #5707 

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

**Comment by DanHeidinga** [comment]
@fengxue-IS Can you test this behaviour on JDK11, 12, & 13?

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk8

**Comment by DanHeidinga** [comment]
Jenkins test extended osx jdk11

**Comment by fengxue-IS** [comment]
I have tried to rename `jdk/lib/libzip.dylib` to `jdk/lib/libzip.jnilib` on Java 11 builds and the library was found and loaded correctly by the JVM.

will test this for Java 12 & 13

**Comment by DanHeidinga** [comment]
Probably worth testing Hotspot's behaviour here as well in case they've dropped this support in newer versions.

**Comment by fengxue-IS** [comment]
Interesting results from testing Hotspot's behaviour:

Looks like Hotspot only allow `.jnilib` library to be loaded using Java System API (`System.loadLibrary()`)
libraries automatically loaded by the JVM (ie. java lib, jvmti agent lib) can only be loaded with `.dylib`

@DanHeidinga do we want to match the RI behaviour?
marking PR WIP while deciding this

**Comment by fengxue-IS** [comment]
looking at the native library loading code, seems like OpenJ9 also permits load library without platform format (`<name>` instead of `lib<name>.ext`)

**Comment by DanHeidinga** [comment]
> @DanHeidinga do we want to match the RI behaviour?

Yes, it doesn't make sense to propagate `.jnilib` into more places than the RI allows it unless there's a compelling reason such as a much cleaner solution by not limiting it.  
`.jnilib` is really a legacy feature that should be supported to ease migration.

**Comment by fengxue-IS** [comment]
Update code to perform the `.jnilib` extension attempt in java level to avoid modifying shared native api for library loading.

@DanHeidinga can you take a look please?

File: jcl/src/java.base/share/classes/java/lang/ClassLoader.java [hunk 1]
```diff
@@ -1896,6 +1896,20 @@ static void loadLibraryWithPath(String libName, ClassLoader loader, String libra
 		}
 	}
 	byte[] message = ClassLoader.loadLibraryWithPath(com.ibm.oti.util.Util.getBytes(libName), loader, libraryPath == null ? null : com.ibm.oti.util.Util.getBytes(libraryPath));
+
+	if (System.internalGetProperties().getProperty("os.name").equals("Mac OS X")) { //$NON-NLS-1$ //$NON-NLS-2$
+		if ((message != null) && (libraryPath != null)) {
+			String legacyPath = libraryPath.replaceAll("/$", "") + "/lib" + libName + ".jnilib"; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$
+
+			try {
+				byte[] message2 = ClassLoader.loadLibraryWithPath(com.ibm.oti.util.Util.getBytes(legacyPath), loader, null);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/ClassLoader.java]: Can you pick a more descriptive name than "message2"?

File: jcl/src/java.base/share/classes/java/lang/ClassLoader.java [hunk 2]
```diff
@@ -1896,6 +1896,20 @@ static void loadLibraryWithPath(String libName, ClassLoader loader, String libra
 		}
 	}
 	byte[] message = ClassLoader.loadLibraryWithPath(com.ibm.oti.util.Util.getBytes(libName), loader, libraryPath == null ? null : com.ibm.oti.util.Util.getBytes(libraryPath));
+
+	if (System.internalGetProperties().getProperty("os.name").equals("Mac OS X")) { //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/ClassLoader.java]: Please add a comment describing why this code is added and is OSX specific

File: jcl/src/java.base/share/classes/java/lang/ClassLoader.java (hunk 1 above)
Comment (fengxue-IS) [reply jcl/src/java.base/share/classes/java/lang/ClassLoader.java]: fixed

File: jcl/src/java.base/share/classes/java/lang/ClassLoader.java (hunk 2 above)
Comment (fengxue-IS) [reply jcl/src/java.base/share/classes/java/lang/ClassLoader.java]: fixed

**Comment by DanHeidinga** [comment]
@fengxue-IS change looks good.  Can you add a test for this as well?  It can be done in a separate PR.

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8

**Comment by fengxue-IS** [comment]
Will open separate PR for test cases

**Pull request merged**